src/index.js

```diff
@@ -82,10 +84,19 @@ class HtmlWebpackEsmodulesPlugin {
         });
       }
       // Write it!
-      fs.writeFileSync(tempFilename, JSON.stringify(newBody));
+      const assetContents = JSON.stringify(newBody);
+      fs.writeFileSync(tempFilename, assetContents);
+
+      // Add the tempfile as an asset so that it will be transformed 
+      // in the PROCESS_ASSETS_STAGE_OPTIMIZE_HASH stage when true
+      // asset hashes are generated
+      const { webpack } = compiler;
+      const { RawSource } = webpack.sources;
+
+      compilation.emitAsset(assetName, new RawSource(assetContents));
```

Comment:
Nice idea about using `emitAsset` thing. But this approach still doesnt work for me (not so sure yet)

Maybe this is because of the way I run multiple builds. It looks like this:

```
const buildConfig = require('./build.production.config');

const modernConfig = buildConfig('modern', ...someOpts);
const legacyConfig = buildConfig('legacy', ...someOpts);

module.exports = [
  modernConfig,
  legacyConfig,
];
```

It feels like two builds act almost in parallel (almost). And the problem (as I think) is that `compilation.emitAsset` behaves in async way, while `fs.writeFileSync` behaves in (surprise!) a synchronous way.

So when it comes to read and parse tempFile, I can get possible situations:
1) if I leave `fs.writeFileSync` in the code, plugin will read unchanged file at iteration 2 (with old hashes)
2) if I delete `fs.writeFileSync` (we have `emitAsset` already), there will be the chance file does not exist yet or it exists but contain nothing

Reply:
Oookay, while I was writing this comment I ran test build for about 10 times and all looks good.
I got temp file reading error only once at first and it was in debugging mode in webstorm. But all "classic" builds via `npm run` command ended successfully. And bundle hashes are correct :tada: 

So maybe it really works :eyes: 


Btw, all builds were with deleted line with `fs.writeFileSync(tempFilename, assetContents);`

Reply:
Great news.  Thanks for testing it!